utopia-remix/app/util/api.server.ts

```diff
@@ -35,25 +36,47 @@ interface HandleRequest {
   params: Params<string>
 }
 
+export type ProjectAccessValidator = (request: Request, params: Params<string>) => Promise<boolean>
+
+export type AccessValidator = ProjectAccessValidator
+
+type Handler = (request: Request, params: Params<string>) => Promise<unknown>
+type ValidatedHandler = {
+  handler: Handler
+  validator: AccessValidator
+}
+
+function toValidatedHandler(descriptor: Handler | ValidatedHandler): ValidatedHandler {
+  if (typeof descriptor === 'function') {
+    return { handler: descriptor, validator: ALLOW }
+  }
+  return descriptor
+}
```

Comment:
I think this is a bit too complex for what we need: imo it would make sense to have the handler data to include a list of extra handlers that run before (and after?), something like:
```ts
handle(args, {
  POST: {
    handler: handleDeleteProject,
    before: [
      authorizeProjectAccess,
      doSomethingElse,
      doAnotherThing
    ],
  }
})
```

and so on. This way we 1) don't need special handler types 2) we immediately think of them as a chain of independent (but sequential!) functions.

Sure this means updating all the existing routes, but imo a bit of repetition is better than a sacrifice in clarity.

Reply:
Yes, I'll update all the routes (I actually think that an explicit `validator` key is important here, since it will enforce always having one)

Reply:
I think this is still overly-specifying responsibilities. I think an even simpler array of handlers would be a more elegant and flexible solution: I really wouldn't want to have to define another key in that type to support a different type of middleware which is functionally exactly the same as a project validator, but has the "unfortunate" property of doing "something else" with a request before going to the actual business logic.

Likewise if I think of _after_ handlers, the same logic would apply.

This is a stub of what I mean (that _also_ does something useful for the future - which is context passing) with a couple of examples: https://gist.github.com/ruggi/32d7e703b739f8b84bae33e4b8a09ba6